cab-token-generator/java/com/google/auth/credentialaccessboundary/ClientSideCredentialAccessBoundaryFactory.java

```diff
@@ -403,6 +467,70 @@ public void run() {
     }
   }
 
+  /**
+   * Serializes a {@link CredentialAccessBoundary} object into Protobuf wire format.
+   */
+  private byte[] serializeCredentialAccessBoundary(
+      CredentialAccessBoundary credentialAccessBoundary) throws IOException {
+    List<AccessBoundaryRule> rules =
+        credentialAccessBoundary.getAccessBoundaryRules();
+    ClientSideAccessBoundary.Builder accessBoundaryBuilder =
+        ClientSideAccessBoundary.newBuilder();
+
+    for (AccessBoundaryRule rule : rules) {
+      ClientSideAccessBoundaryRule.Builder ruleBuilder =
+          accessBoundaryBuilder.addAccessBoundaryRulesBuilder()
+              .addAllAvailablePermissions(rule.getAvailablePermissions())
+              .setAvailableResource(rule.getAvailableResource());
+
+      if (rule.getAvailabilityCondition() != null) {
+        String availabilityCondition =
+            rule.getAvailabilityCondition().getExpression();
+
+        Expr availabilityConditionExpr = this.compileCel(availabilityCondition);
+        ruleBuilder.setCompiledAvailabilityCondition(availabilityConditionExpr);
+      }
```

Comment:
qq, I'm not sure I understand this part. It's creating a ClientSideAccessBoundaryRule.Builder object and adding values to it, but doesn't end up using it again?

Reply:
The `ClientSideAccessBoundaryRule.Builder` is created using `accessBoundaryBuilder.addAccessBoundaryRulesBuilder()` so the builder gets created inside the parent builder `accessBoundaryBuilder`, and the whole rule gets built when `accessBoundaryBuilder.build()` is executed. See: https://protobuf.dev/reference/java/java-generated/#sub-builders

Reply:
Gotcha, that's neat. Thanks!